Jenkins Webhook Integration (#161)
keptn-contrib/artifacthub: jenkins-webhook/1.0.0/README.md

# Jenkins Webhook Integration

> - Using Keptn version 0.10.* to 0.12.x? Use the [Keptn Jenkins Shared Library](https://artifacthub.io/packages/keptn/keptn-integrations/jenkins-library).

Reviewer: ```suggestion
- Using older Keptn version 0.10.x, 0.11.x and 0.12.x: Use the [Keptn Jenkins Shared Library](https://artifacthub.io/packages/keptn/keptn-integrations/jenkins-library).
```

Though in general, the webhook config you provide here most likely still can be used with older Keptn versions

- Using a Keptn version above 0.12.x? Use this integration method.

> Keptn versions over 0.12.x include the webhook service as standard. Use this to trigger Jenkins either asynchronously with no changes required to your Jenkins job. 

Reviewer: ```suggestion
Keptn versions 0.12.x and newer include a webhook capability, which allows you to trigger Jenkins pipelines asynchronously with no changes required to your Jenkins job. 
```

Alternatively trigger Jenkins synchronously with one change to the job.

This video explains both asynchronous and synchronous methods:

> [![Keptn Jenkins Webhook Integration](https://img.youtube.com/vi/ehI23d7s-dY/0.jpg)](https://www.youtube.com/watch?v=ehI23d7s-dY&t=50s)

Reviewer: I highly appreciate making a YouTube video for this! :+1:

Though I would also appreciate if you could add some basic instructions in here, similar to what we have with [Github actions](https://github.com/keptn-contrib/artifacthub/tree/main/githubaction/1.0.0)

# Jenkins Webhook Integration

- Using older Keptn version 0.10.x, 0.11.x and 0.12.x? Use the [Keptn Jenkins Shared Library](https://artifacthub.io/packages/keptn/keptn-integrations/jenkins-library).
- Using a Keptn version above 0.12.x? Use this integration method.

Keptn versions 0.12.x and newer include a webhook capability. Use this to trigger Jenkins either asynchronously with no changes required to your Jenkins job, or synchronously with one change to the job.

## Video Explanation & Demo

This video explains both asynchronous and synchronous methods:

[![Keptn Jenkins Webhook Integration](https://img.youtube.com/vi/ehI23d7s-dY/0.jpg)](https://www.youtube.com/watch?v=ehI23d7s-dY&t=50s)

## Asynchronously Trigger Jenkins

> In the bridge, navigate to a project and go to the Settings > Integrations page.

Reviewer: ```suggestion
In Keptn Bridge, navigate to a project and go to the Settings > Integrations page.
```

> Select the webhook service and click the `+ Add subscription` button.

Reviewer: ```suggestion
Select the webhook-service and click the `+ Add subscription` button.
```

When configuring the webhook, set both `Send started event` and `Send finished event` to `automatically`.

![async started and finished](https://raw.githubusercontent.com/keptn-contrib/artifacthub/main/jenkins-webhook/1.0.0/assets/async-started-finished.jpg)

Craft the webhook to `POST` to the Jenkins `/build` endpoint (see [Jenkins docs](https://www.jenkins.io/doc/book/using/remote-access-api/) for more info).

With these settings, the Keptn `taskName.started` and `taskName.finished` event will be automatically and immediately generated and sent when the webhook is triggered.

The result is that Jenkins is triggered asynchronously.

The task will look like this:

![jenkins async](https://raw.githubusercontent.com/keptn-contrib/artifacthub/main/jenkins-webhook/1.0.0/assets/async-task.jpg)

## Synchronously Trigger Jenkins

> In the bridge, navigate to a project and go to the Settings > Integrations page.

Reviewer: ```suggestion
In Keptn Bridge, navigate to a project and go to the Settings > Integrations page.
```

> Select the webhook service and click the `+ Add subscription` button.

Reviewer: ```suggestion
Select webhook-service and click the `+ Add subscription` button.
```

```
https://myJenkins.com/job/JobName/buildWithParameters?keptnProject={{.data.project}}&keptnService={{.data.service}}&keptnStage={{.data.stage}}&keptnContext={{.shkeptncontext}}&keptnId={{.id}}&keptnTask=<YourTaskName>
```

![sync webhook](https://raw.githubusercontent.com/keptn-contrib/artifacthub/main/jenkins-webhook/1.0.0/assets/sync-webhook.jpg)

Items in double curly brackets are variables which will be pulled from the `taskName.triggered` CloudEvent and replaced at runtime.

These parameters are required so you can build a `taskName.finished` event when the job has finished.

Sending one or more `taskName.status.changed` event(s) **during** execution is also possible. Adjust the example below so the `type` is `sh.keptn.event.<taskName>.status.changed` rather than `sh.keptn.event.<taskName>.finished`.

### Sample Pipeline

```
pipeline {
    agent any

    stages {
        stage('Hello') {
            steps {
                
                echo 'Hello World'
                sleep 10
                
                withCredentials([string(credentialsId: 'keptn-api-token', variable: 'KEPTN_API_TOKEN')]) {
                  sh """

>                     curl -X POST http://myKeptn.com/api/v1/event \

Reviewer: Any chance we can put the KEPTN URL in a variable too? SAme as KEPTN_API_TOKEN? In [Keptn jenkins Library](https://github.com/keptn-sandbox/keptn-jenkins-library#usage) we used an approach where we stored all of it either as a secret or environment variable (I think secret would be better). The variables would then be available in your code as

* `KEPTN_ENDPOINT`
* `KEPTN_BRIDGE` (probably not needed right now)
* `KEPTN_API_TOKEN` - you already have that :+1:

                    -H 'x-token: ${KEPTN_API_TOKEN}' \
                    -H 'Content-Type: application/json' \
                    --data-raw '{
                                "specversion": "1.0",
                                "source": "Jenkins",
                                "shkeptncontext": "${params.keptnContext}",
                                "triggeredid": "${params.keptnId}",
                                "type": "sh.keptn.event.${params.keptnTask}.finished",
                                "contenttype": "application/json",
                                "data": {
                                           "project": "${params.keptnProject}",
                                           "service": "${params.keptnService}",
                                           "stage": "${params.keptnStage}",
                                           "status": "succeeded",

>                                            "result": "pass"

Reviewer: Can we point out that we can also send other statuses and results according to the [Keptn spec](https://github.com/keptn/spec/blob/master/cloudevents.md#data)?
> status: indicates whether the service executing the task was able to perform the task without any unexpected errors. Possible values are succeeded, errored,or unknown
> result: indicates the result of a successful task execution without unexpected problems (i.e. status = succeeded), such as the result of an evaluation, or a test execution. Possible values are pass, warning, or fail

I think this is really a handy way to signal Keptn that your Jenkins pipeline failed.
